examples/vm_scheduling/dqn/vm_rl/vm_dqn.py

```diff
@@ -0,0 +1,141 @@
+import numpy as np
+from typing import Union
+
+import torch
+
+from maro.rl.agent import DQN, DQNConfig
+from maro.rl.model import SimpleMultiHeadModel
+from maro.rl.storage import SimpleStore
+from maro.rl.agent import AbsAgent
+from maro.utils.exception.rl_toolkit_exception import UnrecognizedTask
+
+class VMDQN(AbsAgent):
+    def __init__(
+        self,
+        name: str,
+        model: SimpleMultiHeadModel,
+        config: DQNConfig
+    ):
+        if (config.advantage_mode is not None and
+                (model.task_names is None or set(model.task_names) != {"state_value", "advantage"})):
+            raise UnrecognizedTask(
+                f"Expected model task names 'state_value' and 'advantage' since dueling DQN is used, "
+                f"got {model.task_names}"
+            )
+        super().__init__(
+            name, model, config,
+            experience_pool=SimpleStore(["state", "action", "reward", "next_state", "next_legal_action", "loss"])
+        )
+        self._training_counter = 0
+        self._target_model = model.copy() if model.is_trainable else None
+
+    def choose_action(self, state: np.ndarray, legal_action: np.ndarray) -> Union[int, np.ndarray]:
+        state = torch.from_numpy(state).to(self._device)
+        legal_action = torch.from_numpy(legal_action).to(self._device)
+        is_single = len(state.shape) == 1
+        if is_single:
+            state = state.unsqueeze(dim=0)
+            legal_action = legal_action.unsqueeze(dim=0)
+
+        q_values = self._get_q_values(self._model, state, is_training=False)
+        num_actions = q_values.shape[1]
+
+        greedy_action = self.get_greedy_action(q_values, num_actions, legal_action)
+
+        # No exploration
+        if self._config.epsilon == .0:
+            return greedy_action[0] if is_single else greedy_action
+
+        if is_single:
+            return greedy_action[0] if np.random.random() > self._config.epsilon else self.get_random_action(legal_action[0])
+
+        # batch inference
+        return np.array([
+            act if np.random.random() > self._config.epsilon else self.get_random_action(legal_action[batch_idx])
+            for (batch_idx, act) in enumerate(greedy_action)
+        ])
+
+    def get_greedy_action(self, q_values, num_actions, legal_action):
```

Comment:
mask the value of illegal action to be lower bound value, then use argmax to get the action index with the maximum value, like: 

```
masked_values = q_values * legal_action
greedy_action = argmax(masked_values, dim=...)
```